File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    int func(vector<int> nums, int ind,vector<int> &dp) {
        // base case

        if (ind == 0) {
            return nums[ind];
        }

        if (ind < 0) {
            return 0;
        }

        // take and not take

        // not_take

        if(dp[ind]!=-1){
            return dp[ind];
        }

        int not_take = func(nums, ind - 1,dp);

        int take = INT_MIN;

        // take
        if (ind - 2 >= 0) {
            take = nums[ind] + func(nums, ind - 2,dp);
        }

        return dp[ind] = max(take, not_take);

        // write the DP here
    }
    int rob(vector<int>& nums) {
        int n = nums.size();
        vector<int> dp(n+1,-1);
        

        return func(nums, n - 1,dp);
    }
};

File: 0198-house-robber/0198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        # house robber

        # take and not_take 

        maxi =0
        n = len(nums)
        sum =0
        dp = [-1]*(n+1)

        def func(ind):

            if ind<0:
                return 0
            
            # take and not_take 

            if(dp[ind]!=-1):
                return dp[ind]


            # not take 
            not_take = func(ind-1)

            # sum += nums[ind]

            #take condn 
            take = func(ind-2) + nums[ind]

            dp[ind] = max(take,not_take)

            return dp[ind]




        return func(n-1)



        